### PR TITLE
Improve ASGI request body stream. Fix tests.

### PR DESCRIPTION
**Pull request by hozblok**
In continuation of the [PR](https://github.com/django/channels/pull/1269)

Hi @carltongibson ,

Here are my improvements that can be useful for our work with chunks of the request body.

1) `self._read_started` removed. I found no obstacles to not do this.
2) tests fixed.
3) `RequestBodyWrapper` -> `AsgiRequestIO(io.RawIOBase)`. One method has been implemented: `read()` but we got support `readline`, `close`, `readall` etc according to `io.RawIOBase`. This is quite enough for Django. And an important detail: no need to return exactly `size` bytes in the `read` method. In order not to make `seek` and `truncate` calls I changed `BytesIO` to simple `bytes` object.

I also moved `async_to_sync` (for `receive`) to `__call__` next to the `send` method and made `AsgiRequestIO` purely synchronous. If there are objections, please let me know.

### CODE REVIEW

**Comment by hozblok** [comment]
I will try to put an end to `self._read_started` today.

File: channels/http.py [hunk 1]
```diff
@@ -30,11 +30,10 @@ class AsgiRequest(http.HttpRequest):
     # body and aborts.
     body_receive_timeout = 60
```
Comment (hozblok) [inline channels/http.py:31]: What are we going to do about it? Why do we need `body_receive_timeout`?

**Review by carltongibson — COMMENTED** [review]
Hi @hozblok. 

This is looking really good. (I need to look in daylight, but 👍 so far.) 

>          body_stream = AsgiRequestIO(async_to_sync(receive))

This is interesting. I hadn't thought of putting the `async_to_sync` there. I'm wondering whether we can rely on `receive` always returning control?

**Comment by hozblok** [comment]
Hi @carltongibson 

Placing the `async_to_sync` call here will lead to the same result (the `call_soon_threadsafe` call of the same loop) as putting it inside the `read` method or making it the `read` method decorator. We will get the same `main_event_loop` either explicitly in `async_to_sync` constructor or through the `SyncToAsync.threadlocal` after `await self.handle(body_stream)`.

I'm looking at `receive` implementation in Uvicorn. So the `receive` is not thread-save and contains `asyncio.Event` with `await ...wait()` call. So I cannot find problems with our main loop. But how long will we wait in a separate thread (thread for `await self.handle(body_stream)` call) for future.result() (future in which the `receive` method is wrapped through `main_event_loop.call_soon_threadsafe`)? I don't know, but I hope Daphne or Uvicorn will give us the result after a while or by disconnect. It is for these long operations we move the `handle` to a separate thread. And what do you think about it?

File: channels/http.py (hunk 1 above)
Comment (hozblok) [reply channels/http.py:31]: It used to make sense https://github.com/django/channels/commit/920882f1da3e3c9a9432f2172bf06700bfc6fa48

File: channels/http.py [hunk 2]
```diff
@@ -117,8 +116,8 @@ def __init__(self, scope, body):
             except (ValueError, TypeError):
                 pass
         # Body handling
-        self._stream = body
-
+        self._stream = stream
+        self._read_started = False
```
Comment (hozblok) [inline channels/http.py:120]: I think there should be one `self._read_started = False` in `__init__` as in  [here](https://github.com/django/django/blob/755673e1bca7edb6bee7a958f40d9ae54d85d44c/django/core/handlers/wsgi.py#L96), but not as [here](https://github.com/django/channels/pull/124/files).

**Comment by carltongibson** [comment]
> And what do you think about it?

I need to think about it. 🙂 Lets go with it for now. We can put it in a test app with both Daphne and uvicorn and see that it's well-behaved. (We can always back-out if a problem shows up, and we need to do this anyway before releasing to make sure we didn't introduce an issue.) 

I'm flat-out now until DjangoCon next week, but this (and related issues) is my goal for the sprints there. 

Thank you for your input. It looks really good.

File: tests/test_http.py [hunk 3]
```diff
@@ -26,13 +28,10 @@ def test_basic(self):
         with all optional fields omitted.
         """
 
-        async def recieve():
-            return {}
-
-        body_wrapper = RequestBodyWrapper(recieve)
+        body_stream = AsgiRequestIO(lambda: {})
```
Comment (carltongibson) [inline tests/test_http.py]: I think these calls would be better just inlined now. (No need for the temporary variable.)

File: tests/test_http.py [hunk 4]
```diff
@@ -26,13 +28,10 @@ def test_basic(self):
...
 
         request = AsgiRequest(
-            {"http_version": "1.1", "method": "GET", "path": "/test/"}, body_wrapper
+            {"http_version": "1.1", "method": "GET", "path": "/test/"}, body_stream
```
Comment (carltongibson) [inline tests/test_http.py]: So just `AsgiRequestIO(lambda: {})` here instead of `body_stream`

File: channels/http.py [hunk 5]
```diff
@@ -219,9 +221,8 @@ def __init__(self, scope):
         """
         self.send = async_to_sync(send)
 
-        body = RequestBodyWrapper(receive)
-        await self.handle(body)
-        return
+        body_stream = AsgiRequestIO(async_to_sync(receive))
```
Comment (carltongibson) [inline channels/http.py]: I think just `stream` here.

**Comment by hozblok** [comment]
Hi @carltongibson ,

I committed small edits.

**Comment by hozblok** [comment]
Hi @carltongibson ,

What else can we do for this task?

**Comment by carltongibson** [comment]
Nothing right now. I just need an hour to sit down and think it over a last time. Was hoping to do that at the DjangoCon sprints but didn’t quite get there. 

Thanks for you super input!

**Comment by hozblok** [comment]
You are welcome.
I use this code in my application with uviсorn. So far this works as expected.

---

There is an interesting point with the message about disconnect. We can only get it through `receive`. But we do not do this after receiving the request body. But knowing that a disconnect has occurred could be very useful for interrupting streaming responses in `handle()`:
```
def handle(self, body):
....
for response_message in self.encode_response(response):
    # < --- It would be great to break the 'for' here if disconnected. --->
    self.send(response_message)
```
For example, this could be useful for downloading files from the server through django `StreamingHttpResponse`.
This is not directly related to current work, just information for thought.

**Review by carltongibson — APPROVED** [review]
Hi @hozblok.

Thanks again for this. Since I've been limited on time, I'm going to pull this in as-is and asks others to review. 

👍

**Pull request merged**